src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,167 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    private static class Node<T> {
+        private T item;
+        private Node<T> next;
+        private Node<T> prev;
+
+        public Node(Node<T> prev, T element, Node<T> next) {
+            this.item = element;
+            this.next = next;
+            this.prev = prev;
+        }
+    }
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(null, value, null);
+        if (head == null) {
+            head = newNode;
+        } else {
+            tail.next = newNode;
+            newNode.prev = tail;
+        }
+        tail = newNode;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index > size || index < 0) {
+            throw new IndexOutOfBoundsException("Invalid index " + index);
+        }
+        Node<T> newNode = new Node<>(null, value, null);
+        if (index == 0) {
+            newNode.next = head;
+            if (head != null) {
+                head.prev = newNode;
+            }
+            checkNodeNext(newNode);
+            head = newNode;
+            size++;
+            return;
+        }
+        Node<T> currentNode = getNodeByIndex(index - 1);
+        Node<T> nextNode = currentNode.next;
+        newNode.next = nextNode;
+        checkNodeNext(newNode);
+        if (nextNode != null) {
+            nextNode.prev = newNode;
+        }
+        newNode.prev = currentNode;
+        currentNode.next = newNode;
+        size++;
```

Comment:
Let's distinguish the logic of adding the middle and the end to the beginning more clearly.